v2/component/launcher.go

```diff
@@ -525,8 +525,11 @@ func (l *Launcher) prepareOutputs(ctx context.Context, executorInput *pipeline_s
 			return fmt.Errorf("failed to generate local storage path for output artifact %q with URI %q: %w", name, outputArtifact.Uri, err)
 		}
 
-		if err := os.MkdirAll(filepath.Base(localPath), 0644); err != nil {
-			return fmt.Errorf("unable to create directory %q for output artifact %q: %w", localPath, name, err)
+		if err := os.MkdirAll(filepath.Dir(localPath), 0644); err != nil {
+			return fmt.Errorf("unable to create directory %q for output artifact %q: %w", filepath.Dir(localPath), name, err)
+		}
+		if _, err:= os.Create(localPath); err != nil {
```

Comment:
I'm not sure whether we should create a file path for user? Are we able to know whether this `localPath` is a directory or a file path? For example, I found in this example https://github.com/kubeflow/pipelines/blob/bb967e9e1ea8acb9cc198bc1c627bae1afb78078/samples/core/output_a_directory/output_a_directory.py#L32, it can be a directory. But in the metrics_visualization example I created in this PR, the localPath is a filePath(not a directory).

Reply:
In both cases, we should create parent folder of the file path. If component wants to output a directory artifact it's their own responsibility to create that folder

Reply:
Thanks for the fix!

Reply:
@Bobgy Just to clarify, for the example of the metrics visualization, the launcher is only responsible for creating the parent folder of the file path(for example /dirA/dirB, the component itself is responsible for creating the file path(for example /dirA/dirB/fileC), right?

Reply:
>If component wants to output a directory artifact it's their own responsibility to create that folder

+1

>the component itself is responsible for creating the file path(for example /dirA/dirB/fileC), right?

Yes.

>we should create parent folder of the file path
>the launcher is only responsible for creating the parent folder of the file path(for example /dirA/dirB

When using custom containers, KFP v1 and Cloud IR backends do not create even the parent directory (since they cannot easily manipulate filesystem inside the container). So, in general it's also the responsibility of the component code to create the parent directory.
On exception is KFP's Lightweight python components where the wrapper pre-creates parent directories. The directory creation code is part of the generated component program code. So in some sense the principle still holds - the component code is responsible for the creation of the parent directory.

Reply:
Looks like this parent folder creation is an intended UX improvement https://github.com/kubeflow/pipelines/pull/5458#discussion_r613783441

Reply:
After I create the parent folder of filePath, it still fails with `no such file or directory`. Looks like there is something wrong with sdk side or component side @chensun

Reply:
Regarding the error of no such file or directory, talked with @chensun  and @neuromage , metrics is a special artifact that we don't write any payload but only metadata. We willdrop the expectation that there must be a file. In this case, we can just ignore the missing payload. Maybe log a warning. But allow the component to succeed rather than fail when upload/download does not work. The URI is essentially a dummy URI in this case.